### PR TITLE
Tutorial] 7. tensor flow amp

### PR DESCRIPTION
**Pull request by cxheng315**
#61

### CODE REVIEW

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 1]
```diff
@@ -1 +1,305 @@
...
+
+```bash=
+$ nvidia-smi --loop=10
+``` 
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_cf31d97e7819afb3bcdc7f8137c2ae5b.png)
+
+
+
+After the model training is completed, and it can process about 393.49 images per second:
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:169]: [SUG]

The model training completed, and It could process about 393.49 images per second:

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 2]
```diff
@@ -1 +1,305 @@
...
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_9a1072b7574cc7dc6abcee91ef520147.png)
+
+
+
+###  Group 2. Batch Size 256
+
+> Unactivate AMP | Double precision ｜ Batch size = 256
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:177]: [SUG]

Unactivate AMP > AMP disabled

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 3]
```diff
@@ -1 +1,305 @@
...
+
+
+Processing about 405.73 images per second, showing 1.03 times better performance than the control group:
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_2cf72e347cfc8a17634c2ed9c1f4fa0a.png)
+
+
+
+
+###  Group 3. Activate AMP
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:201]: [SUG]

Activate AMP > AMP enabled

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 4]
```diff
@@ -1 +1,305 @@
...
+
+> Activate AMP | Mixed precision (single and half precision)｜Batch size = 256
+
+Enter the following command to set the environment variable to activate AMP, and perform computings:
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:205]: [SUG]

1. activate AMP > enable AMP
2. computings > computing

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 5]
```diff
@@ -1 +1,305 @@
...
+
+
+
+Processing about 1308.37 images per second, showing 3.33 times better performance than the control group:
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_b4251f877ae84fc3c466ac5330d8b041.png)
+
+
+
+###  Group 4. Activate XLA
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:229]: [SUG]

Activate XLA > XLA enabled

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 6]
```diff
@@ -1 +1,305 @@
...
+
+> Activate AMP,XLA | Mixed precision (single and half precision)｜Batch size = 256
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:231]: AMP & XLA enabled

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 7]
```diff
@@ -1 +1,305 @@
...
+
+
+
+Processing about 1309.21 images per second, showing 3.33 times better performance than the control group:
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_926a8e53f6578ecae4cf7d19fbb6e81f.png)
+
+
+
+###  Group 5. Double the Batch Size
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:253]: [SUG]
Double the Batch Size > Double the batch size

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 8]
```diff
@@ -1 +1,305 @@
...
+
+> Activate AMP,XLA | Mixed precision (single and half precision) ｜Batch size = 512
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:255]: [SUG]

Activate AMP,XLA  > AMP & XLA enabled

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 9]
```diff
@@ -1 +1,305 @@
...
+
+
+:::spoiler ResNet-50 v1.5 Benchmark performance analysis computings example
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:258]: [SUG]

computings example > example

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 10]
```diff
@@ -1 +1,305 @@
...
+::: info 
+You can directly analyze the performance of ResNet-50 through the following commands, the running time is about 3 minutes 
+
+`wget -q -O - http://bit.ly/TWCC_CCS_AMP-XLA | bash`
+
+:::
+
+
+
+AMP reduces the usage of GPU memory, so we doubled the Batch Size in an attempt to increase the execution performance. Enter the command directly in the command line as follows:
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:268]: [SUG]

Batch Size > batch size

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 11]
```diff
@@ -1 +1,305 @@
...
+
+Processing about 1361.00 images per second, showing 3.46 times better performance than the control group:
+
+![](https://cos.twcc.ai/SYS-MANUAL/uploads/upload_7da55c796423eb1359f95ecf7ab4db35.png)
+
+
+
+### Performance comparison
+
+Summarizing the results of each group of image processing data (img/sec, GPU memory usage), the performance comparison is shown in the following chart. Compared with the control group, the performance of turning on AMP acceleration is quite significant, reducing GPU memory usage and shortening training computing time:
```
Comment (VigaWei) [inline CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:290]: [SUG]

turning on > enabling

**Review by VigaWei — COMMENTED** [review]
Thx! Please revise the doc according to the comments.

File: CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md [hunk 12]
```diff
@@ -1 +1,305 @@
...
+- [SSH connection to enter container](#SSH-connection-to-enter-container)
+- [Activate AMP function](#Activate-AMP-function) 
+    - [Environment variable setting method](#Environment-variable-setting-method)
+    - [Example of MNIST handwriting recognition program](#Example-of-MNIST-handwriting-recognition-program)
+- [Benchmark performance analysis:ResNet-50 v1.5](#Benchmark-performance-analysis:ResNet-50-v1.5)
+
+
+## Introduction to AMP
+
+Traditional high-performance computings using Double Precision computings to ensure the convergence feature of numerical algorithms (for example, atmospheric simulation). Double precision computing (FP32) allows each value to occupy 32 bits and can reach the accuracy of 16 fields, and this computing requires a lot of memory space.
```
[the remark this answers is not in the document]
Comment (VigaWei) [reply CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:25]: @twsc-aug help, please!

Not sure of this:  "reach the accuracy of 16 fields"

[the remark this answers is not in the document]
Comment (twsc-aug) [reply CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:25]: However, Double precision computing (FP32) requires a lot of memory space due to it uses 32 bits to represent 16 digits of floating point.
然而雙精準度運算 (FP32) 讓每個數值佔據 32 個位元，以可達到 16 位小數點浮點數的精準度，而此種計算需要大量記憶體空間。

[the remark this answers is not in the document]
Comment (twsc-aug) [reply CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:25]: @VigaWei

[the remark this answers is not in the document]
Comment (VigaWei) [reply CCS/3. Tutorials/7. 開啟 TensorFlow 自動混合精度運算與執行效能分析.md:25]: 小改: However, Double precision computing (FP32) requires a lot of memory space since it uses 32 bits to represent 16 digits of floating point.
